apps/src/music/progress/ProgressManager.ts

```diff
@@ -0,0 +1,129 @@
+// This file contains a generic ProgressManager which any lab can include,
+// if it wants to make progress without reloading the page.
+
+// Abstract class that validates a set of conditions. How
+// the validation works is up to the implementor.
+export abstract class Validator {
+  abstract shouldCheckConditions(): boolean;
+  abstract checkConditions(): void;
+  abstract conditionsMet(conditions: string[]): boolean;
+  abstract clear(): void;
+}
+
+// A validation inside the progression.
+interface Validation {
+  conditions: string[];
+  message: string;
+  next: boolean;
+}
+
+// The definition of the progression.
+interface Progression extends Object {
+  panels: {
+    text: string;
+    toolbox: {
+      [key: string]: string;
+    };
+    sounds: {
+      [key: string]: string;
+    };
+    validations: Validation[];
+  }[];
+}
+
+// The current progress state.
+interface ProgressState {
+  step: number;
+  satisfied: boolean;
+  message?: string | null;
```

Comment:
nit: this can just be optional (`message?: string`) unless we specifically want to default to null, in which case it should probably be a required `message: string | null` - as written this is effectively `message: string | null | undefined`

Reply:
Oh good catch.